### PR TITLE
Set frame rate of video extracting

### PR DESCRIPTION
**Pull request by zliang7**
This value means how many image files generated per second.
To reduce the number of image files, the value must be less than
the video FPS.

### CODE REVIEW

**Comment by zliang7** [comment]
This PR is trying to fix #194. To reduce the number of image files, this PR use frame rate value(FPS) Instead of using step value of frame.
@nmanovic , PTAL

File: cvat/apps/engine/task.py [hunk 1]
```diff
@@ -272,13 +272,17 @@ def rq_handler(job, exc_type, exc_value, traceback):
...
         translated_quality = 96 - compress_quality
         translated_quality = round((((translated_quality - 1) * (31 - 2)) / (95 - 1)) + 2)
         self.output = tempfile.mkdtemp(prefix='cvat-', suffix='.data')
         target_path = os.path.join(self.output, '%d.jpg')
-        output_opts = '-start_number 0 -b:v 10000k -vsync 0 -an -y -q:v ' + str(translated_quality)
+        output_opts = '-start_number 0 -b:v 10000k -an -y -q:v ' + str(translated_quality)
+        if frame_rate > 0:
+            output_opts += ' -r ' + str(frame_rate)
+        else:
+            output_opts += ' -vsync 0'
```
Comment (nmanovic) [inline cvat/apps/engine/task.py]: Option `-vsync 0` is very important. It allows to generate non-duplicated frames. -r can lead to duplicated frames and in general it is bad
- need to annotate the same frame twice
- there is no correspondence between frame and video. What does frame #3 mean? Is it duplication of frame #2? Is it an unique frame?

**Comment by nmanovic** [comment]
nm/rest_api should replace develop very soon. Could you please submit your changes into nm/rest_api branch?

File: cvat/apps/engine/task.py [hunk 2]
```diff
@@ -272,13 +272,17 @@ def rq_handler(job, exc_type, exc_value, traceback):
 ############################# Internal implementation for server API
 
 class _FrameExtractor:
-    def __init__(self, source_path, compress_quality, flip_flag=False):
+    def __init__(self, source_path, compress_quality, flip_flag=False, frame_rate=0):
```
Comment (nmanovic) [inline cvat/apps/engine/task.py]: The main problem with the approach is you will get annotation file which doesn't correspond to the video file. For example, originally you have a video file with 30FPS. If someone annotates the whole video file we will get `<frame000000>, <frame000001>, ...` inside the annotation file for every frame. Now you say that we can provide "custom" FPS. You will get another `<frame000000>, <frame000001>, ...` which don't correspond to the original video file. How to much if you don't store extra parameters (e.g. custom frame rate) inside the annotation file?

File: cvat/apps/engine/task.py (hunk 1 above)
Comment (zliang7) [reply cvat/apps/engine/task.py]: I have tried to keep the vsync options. But if both -vsync and -r options are specified, ffmpeg reports error:
"Using -vsync 0 and -r can produce invalid output files"

**Comment by nmanovic** [comment]
Let me clarify my vision for the feature:
- Need several extra fields:
  * start frame. It should be 0 by default, >=0, less than stop frame. It will be good if you can specify it as frame number of as time
  * stop frame. It should be empty by default (corresponds to the end), >=0, more than start frame. It will be good if you can specify it as frame number of as time
  * step. It should be empty by default (corresponds to the minimum possible step), >0. It will be good if you can specify it as frame number of as time

- Inside dump file you should have exact correspondence between the video file and dumped frames. For example, if you annotate only 1 frame from 30 frames you should be `<frame000000>, <frame000029>, ...` inside the annotation file.

- As I said previously the patch should be submitted into nm/rest_api branch. It will be merged into develop branch very soon.

File: cvat/apps/engine/task.py (hunk 1 above)
Comment (nmanovic) [reply cvat/apps/engine/task.py]: @zliang7, this is why '-r' option is not good here. See my thoughts below.

**Comment by nmanovic** [comment]
>   * start frame. It should be 0 by default, >=0, less than stop frame. It will be good if you can specify it as frame number or as time

May be for unification the default value for `start frame` should be also empty.

**Comment by nmanovic** [comment]
One more important note. The same approach can work not only for a video file. It can work for a set of images or an archive. All these parameters above should be as a part of rest api and go to DB as well.

File: cvat/apps/engine/task.py (hunk 2 above)
Comment (zliang7) [reply cvat/apps/engine/task.py]: I'm not familiar with the internals of backend. Is there any difference between upload a video directly with upload all frame images pre-extracted locally?

Comment (nmanovic) [reply cvat/apps/engine/task.py]: @zliang7, if you still have questions it is better to organize a meeting and discuss.

**Comment by nmanovic** [comment]
@zliang7 , should we expect a new patch for the issue or just close the PR?

**Comment by zliang7** [comment]
@nmanovic , it's OK to close this PR. I will try to submit again after your restful api PR get merged. But before that, I need some time to understand your huge change to rebase/change this one.

I still have some questions about your thoughts. Instead of setup a meeting to discuss, I prefer to discuss in the issue page. So everyone can see it and join the discussion, since 
cvat is an opensource project. Please see my question at #194

**Comment by nmanovic** [comment]
OK. Let's make one more iteration with the PR. I will close it for now.

**Pull request closed without merging**